Add Login Screen handler

BitScorpio opened the pull request.

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/blocking/BlockingScript.java:
```diff
@@ -0,0 +1,60 @@
+            b = background.getBlue(),
+            a = background.getAlpha();
+
+        if (increment) {
+            if (a == 150) {
+                increment = false;
+                background = new Color(r, g, b, a - 1);
+            } else {
+                background = new Color(r, g, b, a + 1);
+            }
+        } else {
+            if (a == 100) {
+                increment = true;
+                background = new Color(r, g, b, a + 1);
+            } else {
+                background = new Color(r, g, b, a - 1);
+            }
+        }
```
```suggestion
        increment = (increment && a < 150) ||  a <= 100;
        background = new Color(r, g, b, a - increment ? -1 : 1);
```

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/blocking/BlockingScript.java:
```diff
@@ -0,0 +1,60 @@
+import org.rspeer.game.event.RenderEvent;
+import org.rspeer.game.event.listener.RenderListener;
+
+public abstract class BlockingScript implements RenderListener {
```
Not sure I like the script name, I'd prefer it be named BlockingGameEvent or GameEventHandler

Jasper-ketelaar replied:
Maybe even make an interface for blocking to allow it to be more generic

BitScorpio replied:
This will cause confusion with dispatched game events

Jasper-ketelaar replied:
Right.. Any other suggestions? GameSituationHandler or GameSituationResponse perhaps? I'd just prefer to not use the word Script here.

Lezor91 replied:
Here's my proposition(s):
- extend Script instead, map them with a GameSituation enum (Also makes it easier to provide set/remove methods compared to this ArrayList we have)
- Change the way validate is currently called, it will be terrible when we have a lot of situation handlers being checked every loop (Known issue in rsp2)
- Change the execution process so that it temporarily changes the active script?
- validate can be provided in GameSituation perhaps?

BitScorpio replied:
**Implemented:**
> extend Script instead

**Pending discussion:**
>  map them with a GameSituation enum (Also makes it easier to provide set/remove methods compared to this ArrayList we have)

> validate can be provided in GameSituation perhaps?

Not sure if this is the best thing to do if we're going to support a custom **GameSituationHandler** from within the script. We need to define the validation rules for each **GameSituationHandler** depending on what it's intended to handle.

> Change the way validate is currently called, it will be terrible when we have a lot of situation handlers being checked every loop (Known issue in rsp2)

I can't think of a way to make sure the validation method is called for every **GameSituationHandler** we have other than looping through all of the currently existing ones. Maybe we could implement a time restriction on how frequently each validation method should be called (Defined inside the **GameSituationHandler** instance) if we're concerned about constantly running expensive checks (Checking for a Genie random event for instance)

> Change the execution process so that it temporarily changes the active script?

It shouldn't really change the active script, it'd just call the loop method directly. I can already foresee many problems occurring when trying to stop/reload the current script if we actually replace the active script variable and it'll require us to complicate the entire process.

d-o-g replied:
> It shouldn't really change the active script, it'd just call the loop method directly.
calling loop directly removes the point of extending script, we want to gain the benefits of using the framework - processing other logic such as onstarts and listeners. the already active script wouldnt need to be stopped/reloaded, it can just be put on hold

> I can't think of a way to make sure the validation method is called for every GameSituationHandler we have other than looping through all of the currently existing ones.
perhaps event listener based triggers will work, but that can probably wait until the script engine is added

> Maybe we could implement a time restriction on how frequently each validation method should be called
this can probably work as a temporary solution

> Not sure if this is the best thing to do if we're going to support a custom GameSituationHandler from within the script. We need to define the validation rules for each GameSituationHandler depending on what it's intended to handle.
a map of gamesituation -> gamesituationhandler works too, but we cant benefit from using enums because that removes the option to add custom situations (although this would be trivial, perhaps we can ignore it)

yes i do read from bottom to top so what

BitScorpio replied:
> calling loop directly removes the point of extending script, we want to gain the benefits of using the framework - processing other logic such as onstarts and listeners. the already active script wouldnt need to be stopped/reloaded, it can just be put on hold

I'll introduce an **activeSituation** variable to hold reference to the currently active **GameSituationHandler**.


> a map of gamesituation -> gamesituationhandler works too, but we cant benefit from using enums because that removes the option to add custom situations (although this would be trivial, perhaps we can ignore it)

I'd wait until we decide which approach to take for triggering the GameSituationHandlers

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/account/Account.java:
```diff
@@ -0,0 +1,20 @@
+package org.rspeer.game.script.account;
+
+public class Account {
```
GameAccount to prevent confusion

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/Script.java:
```diff
@@ -41,32 +49,76 @@ public final void setState(State state) {
+
+    @Override
+    public final void run() {
+        registerBlockingScript(new LoginScreen(this));
```
Move this to its own method

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/Script.java:
```diff
@@ -41,32 +49,76 @@ public final void setState(State state) {
+                    if (blockingScript.validate()) {
+                        sleep = blockingScript.execute();
+                        Time.sleep(sleep);
+                        break;
```
Its not blocking if you only break it, needs a return.

Jasper-ketelaar replied:
Oh, that's a bit confusing

BitScorpio replied:
I've changed the implementation, no longer need to run such check anyway, marking this as resolved.

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/Script.java:
```diff
@@ -1,15 +1,23 @@
 //TODO make this better, add random handling (login screen, welcome screen etc)
 public abstract class Script implements Runnable {
 
+    private static Account account;
```
Is it the scripts responsibility to manage the logged in account? Don't like this

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/Script.java:
```diff
@@ -41,32 +49,76 @@ public final void setState(State state) {
-                setState(State.STOPPED);
-                return;
-            }
+    public static void setAccount(Account account) {
```
Shouldn't be the scripts responsibility.

Jasper-ketelaar commented on scripting/src/main/java/org/rspeer/game/script/blocking/BlockingScript.java:
```diff
@@ -0,0 +1,60 @@
+            Graphics2D g = (Graphics2D) e.getSource();
+            g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_LCD_HRGB);
+
+            adjustBackround();
```
Name

Lezor91 commented:
Script can still access account via non static getter? Also there's a GameAccount class in the API already. 

In rsp2 the account is a non static field with getter/setter in Script, and it is set to the logged in user upon startup (if the script is started logged in).

As far as the login code itself goes, there's a much simpler way of doing it. 
`
client.setUsername(user)
client.setPassword(pw)
client.setTotp(auth)
client.setGameState(20)
`
With this approach you do not have to concern yourself with emulating keyboard functionality or cycling login states individually. This is probably more future-proof too, recent changes in the engine indicate that the login screen may be changed, however, it is less likely that they will change the underlying internal code so this approach is probably more safe anyhow.

MaddoxDevelopment commented:
There is a lot more states that this should handle by default before being accepted imo. It doesn't have to be 100% perfect but you should at-least make it handle such things as error connecting to server, switching to non members world if non mem, stopping on ban, too many login tries, etc.

Lezor91 commented:
This iteration wouldn't support it, as we do not have a response listener in rsp3 yet.

MaddoxDevelopment commented:
> This iteration wouldn't support it, as we do not have a response listener in rsp3 yet.

There is one in EventMediator though? You don't technically need it either, could just check the menu text & login step each loop.

Lezor91 commented:
Ah my bad, forgot that I added it

BitScorpio commented:
> There is a lot more states that this should handle by default before being accepted imo. It doesn't have to be 100% perfect but you should at-least make it handle such things as error connecting to server, switching to non members world if non mem, stopping on ban, too many login tries, etc.

Not so sure if we want to switch to non-members world, it could break scripts intended to run on P2P in case the membership runs out mid-execution.

Don't really need to add explicit handling for error connecting to server. It should always attempt to login if there's nothing implemented for specific things.

Apart from that, I'll work on as many situations as possible. I already have most of them handled in a custom rsp2 handler so as you said, shouldn't be a problem porting them.

d-o-g commented:
rsp2 login handler offers options such as stopping script on certain responses - updates, bans, locks by default with the option to add more responses to stop on

```            if (clientUsername.length() == 0) {
                Game.getClient().setUsername(scriptUsername);
            } else if (clientPassword.length() == 0) {
                Game.getClient().setPassword(scriptPassword);
            } else if (!clientUsername.equals(scriptUsername)) {
                Game.getClient().setUsername("");
            } else if (!clientPassword.equals(scriptPassword)) {
                Game.getClient().setPassword("");
            } else {
                Game.getClient().setGameState(Game.STATE_LOGGING_IN);
            }```

this is pretty bad, you dont need to check what the existing field values contain, just set them directly. the else branches also means it will only visit a single branch per iteration. granted, the default sleep is 1 ms so it's not a big deal but itd still be cleaner and its completely safe to take said approach

d-o-g commented:
how the fuck do codet ahgs work on here

BitScorpio commented:
> 
> 
> how the fuck do codet ahgs work on here

I use preview before commenting * Eye-rolling *

d-o-g commented:
wish u would preview ur ganks too

BitScorpio commented:
**The following changes have been made:**
- **GameSituationHandler** is now treated as an alternative Script that starts/stops based on validation
- Removed all **ScriptChangeEvent** dispatching since it is already handled by **Script.setState()**
- Fixed a bug with script reloading not working properly (Probably caused by my **ScriptThread** implementation)

**The following is not implemented yet:**
- Support for more login states (Currently supports LOGGED_OUT & INVALID_CREDENTIALS)

**The following was discussed but not implemented:**
- Holding **GameSituationHandler**s in a different datatype (Currently List). My main concern is the effect this would have on supporting custom **GameSituationHandler**. Someone out there might want to, for example, implement random events solving. I want to wait on this until events-based triggering is looked into as an alternative to validation-based triggering.

Lezor91 commented:
Calling validate in paint is a bad idea, keep in mind that paint calls are executed every game frame which means validate will be called very frequently.

BitScorpio commented:
> 
> 
> Calling validate in paint is a bad idea, keep in mind that paint calls are executed every game frame which means validate will be called very frequently.

Will take care of that

Lezor91 commented:
I have the idea in mind to expand the script controller to better support situations like this where we are running "multiple" scripts at the same time, possibly controlled by a priority system and/or activation triggers, it should make this implementation a bit nicer

BitScorpio commented:
> 
> 
> I have the idea in mind to expand the script controller to better support situations like this where we are running "multiple" scripts at the same time, possibly controlled by a priority system and/or activation triggers, it should make this implementation a bit nicer

Yeah I don't like this implementation at all neither, I'll close this PR and come up with something better specially that we've changed the way events get dispatched.

Lezor91 commented:
Yeah sounds good, it's not an implementation issue but rather an issue at the higher level, solving it further up would mean less issues for the future
